jupyterhub/scopes.py

```diff
@@ -646,68 +698,112 @@ def _resolve_requested_scopes(requested_scopes, have_scopes, user, client, db):
     return (allowed_scopes, disallowed_scopes)
 
 
-def _needs_scope_expansion(filter_, filter_value, sub_scope):
+def _needs_group_expansion(filter_, filter_value, sub_scope):
     """
     Check if there is a requirements to expand the `group` scope to individual `user` scopes.
     Assumptions:
     filter_ != Scope.ALL
     """
-    if not (filter_ == 'user' and 'group' in sub_scope):
+    if not (filter_ in {'user', 'server'} and 'group' in sub_scope):
         return False
-    if 'user' in sub_scope:
-        return filter_value not in sub_scope['user']
+    if filter_ in sub_scope:
+        return filter_value not in sub_scope[filter_]
     else:
         return True
 
 
-def _check_user_in_expanded_scope(handler, user_name, scope_group_names):
-    """Check if username is present in set of allowed groups"""
-    user = handler.find_user(user_name)
-    if user is None:
-        raise web.HTTPError(404, "No access to resources or resources not found")
-    group_names = {group.name for group in user.groups}
-    return bool(set(scope_group_names) & group_names)
+def _has_scope_key(scope, have_scopes, *, post_filter=False, db=None):
+    """Cache key for has_scope"""
+    if isinstance(have_scopes, dict):
+        have_scopes = FrozenDict(have_scopes)
+    else:
+        have_scopes = frozenset(have_scopes)
+    return (scope, have_scopes, post_filter)
+
 
+@lru_cache_key(_has_scope_key)
+def has_scope(scope, have_scopes, *, post_filter=False, db=None):
```

Comment:
Since this utility is important and used often in the new code, I've promoted checking for a scope to a simple public API